server/chat-plugins/net-filters.ts

```diff
@@ -0,0 +1,217 @@
+/**
+ * Neural net chatfilters.
+ * These are in a separate file so that they don't crash the other filters.
+ * (issues with globals, etc)
+ * by Mia.
+ * @author mia-pi-git
+ */
+
+import {QueryProcessManager, QueryProcessWrapper} from '../../lib/process-manager';
+import {FS} from '../../lib/fs';
+import {Utils} from '../../lib/utils';
+import {Config} from '../config-loader';
+
+const PATH = "config/chat-plugins/net.json";
+const MAX_PROCESSES = Config.netfilterprocesses || 1;
+const PM_TIMEOUT = 2 * 60 * 60 * 1000; // training can be _really_ slow
+const WHITELIST = ["mia"];
+
+interface NetQuery {
+	data: string | TrainingLine[];
+	type: "run" | "train" | "save" | "load";
+	options?: AnyObject;
+}
+
+interface TrainingLine {
+	input: string;
+	// "|flag" (bad) "|ok" (good)
+	output: string;
+}
+
+export class NeuralNetChecker {
+	model: import('brain.js').recurrent.LSTM | null;
+	constructor(path?: string) {
+		try {
+			this.model = new (require('brain.js').recurrent.LSTM)();
+		} catch (e) {
+			this.model = null;
+		}
+		if (path) this.load(path);
+	}
+	async train(data: TrainingLine[], iterations?: number) {
+		// 100 has good perf but is still effective
+		if (!iterations) iterations = 100;
+		const now = Date.now();
+		await FS(PATH).copyFile(PATH + '.backup');
+		this.model?.train(data, {iterations});
+		this.save();
+		return Date.now() - now; // time data is helpful for training
+	}
+	save(path = PATH) {
+		if (!this.model) return {};
+		const state = this.model.toJSON();
+		FS(path).writeUpdate(() => JSON.stringify(state));
+		return state;
+	}
+	load(path: string) {
+		if (!FS(path).existsSync()) return;
+		const data = JSON.parse(FS(path).readSync());
+		this.model?.fromJSON(data);
+	}
+	run(data: string) {
+		let result = '';
+		if (!this.model) return result;
+		try {
+			result = this.model.run(data);
+		} catch (e) {}
+		// usually means someone didn't train it, carry on
+		// acceptable to drop since training is very slow
+		return result;
+	}
+	static trainingProcesses: QueryProcessWrapper[] = [];
+	static async train(data: TrainingLine[]) {
+		const process = PM.createProcess();
+		this.trainingProcesses.push(process);
+		// do the training in a separate process (which we never add to PM.processes so it doesn't get pinged for other requests)
+		const result = await process.query({type: 'train', data});
+		// kill process
+		this.trainingProcesses.splice(this.trainingProcesses.indexOf(process), 1);
+		process.destroy();
+		// load it into the main process that we're querying
+		await PM.query({type: 'load', data: PATH});
+		return result;
```

Comment:
I explained in DMs, but there's no reason for you to use a process manager for this. This is exactly what `child_process` is for.

Reply:
Also: this is already a subprocess! PM was never written for sub-subprocesses.

Reply:
No? This is run from the command in the main process.

Reply:
Oh, I see. Still, this is what `child_process` is for.

Reply:
Updated to handle new refactor!